src/signalrclient/connection_impl.cpp

```diff
@@ -306,44 +332,95 @@ namespace signalr
                 }
             });
 
-        pplx::create_task([connect_request_tce, disconnect_cts, weak_connection]()
+        std::thread([disconnect_cts, connect_request_done, connect_request_lock, callback, weak_connection]()
         {
-            // TODO? std::this_thread::sleep_for(std::chrono::milliseconds(negotiation_response.transport_connect_timeout));
-            std::this_thread::sleep_for(std::chrono::milliseconds(5000));
+            disconnect_cts->wait(5000);
 
             // if the disconnect_cts is canceled it means that the connection has been stopped or went out of scope in
             // which case we should not throw due to timeout. Instead we need to set the tce prevent the task that is
             // using this tce from hanging indifinitely. (This will eventually result in throwing the pplx::task_canceled
             // exception to the user since this is what we do in the start() function if disconnect_cts is tripped).
-            if (disconnect_cts.get_token().is_canceled())
+            if (disconnect_cts->is_canceled())
             {
-                connect_request_tce.set();
+                bool run_callback = false;
```

Comment:
nit: you can collapse these methods to get run_callback before checking disconnect_cts->is_canceled().